### PR TITLE
OHRM5X-1385: Develop performance tracker logs BE

### PR DESCRIPTION
**Pull request by SharukaNP**

### CODE REVIEW

File: installer/Migration/V5_1_0/permission/api.yaml [hunk 1]
```diff
@@ -113,6 +113,19 @@ apiv2_performance_review:
...
+  description: 'API-v2 Performance - My Tacker Logs'
+  api: OrangeHRM\Performance\Api\PerformanceTrackerLogAPI
+  module: performance
+  allowed:
+    read: true
+    create: true
+    update: true
+    delete: true
+  permissions:
+    - { role: Admin, permission: { read: true, create: true, update: true, delete: true, self: true } }
```
[the remark this answers is not in the document]
Comment (RajithaKumara) [reply installer/Migration/V5_1_0/permission/api.yaml:127]:
```suggestion
```
remove this permission

File: installer/Migration/V5_1_0/permission/api.yaml [hunk 2]
```diff
@@ -113,6 +113,20 @@ apiv2_performance_review:
   permissions:
     - { role: Admin, permission: { read: true, create: true, update: true, delete: true, self: false } }
 
+apiv2_performance_my_tracker_logs:
+  description: 'API-v2 Performance - My Tacker Logs'
```
Comment (RajithaKumara) [inline installer/Migration/V5_1_0/permission/api.yaml]:
```suggestion
  description: 'API-v2 Performance - Tacker Logs'
```

File: src/plugins/orangehrmPerformancePlugin/Api/Model/DetailedPerformanceReviewModel.php [hunk 3]
```diff
@@ -19,49 +19,56 @@
...
+use OrangeHRM\Core\Api\V2\Serializer\ModelTrait;
 use OrangeHRM\Core\Api\V2\Serializer\Normalizable;
-use OrangeHRM\Core\Traits\Service\NormalizerServiceTrait;
 use OrangeHRM\Entity\PerformanceReview;
 use OrangeHRM\Performance\Traits\Service\PerformanceReviewServiceTrait;
 
 class DetailedPerformanceReviewModel implements Normalizable
 {
-    use NormalizerServiceTrait;
+    use ModelTrait {ModelTrait::toArray as modelToArray;}
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPerformancePlugin/Api/Model/DetailedPerformanceReviewModel.php:29]:
```suggestion
    use ModelTrait {ModelTrait::toArray as entitylToArray;}
```
shall we go like this.

File: src/plugins/orangehrmPerformancePlugin/Api/Model/PerformanceTrackerLogModel.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                ['reviewer', 'empNumber'],
+                ['reviewer', 'lastName'],
+                ['reviewer', 'firstName'],
+                ['reviewer', 'middleName'],
+                ['reviewer', 'terminationId'],
+            ]
+        );
+    }
+
+    public function toArray(): array
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPerformancePlugin/Api/Model/PerformanceTrackerLogModel.php:69]: better if you can add doc comment

File: src/plugins/orangehrmPerformancePlugin/Api/PerformanceTrackerLogAPI.php [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+            ->getPerformanceTrackerLogsRateCount(PerformanceTrackerLog::POSITIVE_RATING, $performanceTrackerLogParamHolder->getTrackerId());
+        $performanceTrackerLogsNegativeRatingCount = $this->getPerformanceTrackerLogService()
+            ->getPerformanceTrackerLogDao()
+            ->getPerformanceTrackerLogsRateCount(PerformanceTrackerLog::NEGATIVE_RATING, $performanceTrackerLogParamHolder->getTrackerId());
+        return new EndpointCollectionResult(
+            PerformanceTrackerLogModel::class,
+            $performanceTrackerLogs,
+            new ParameterBag(
+                [
+                    CommonParams::PARAMETER_TOTAL => $performanceTrackerLogCount,
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPerformancePlugin/Api/PerformanceTrackerLogAPI.php:92]:
```suggestion
                    CommonParams::PARAMETER_TOTAL => $performanceTrackerLogsPositiveAchievementCount, + $performanceTrackerLogsNegativeAchievementCount,
```
just an idea.

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmPerformancePlugin/entity/PerformanceTrackerLog.php [hunk 6]
```diff
@@ -21,13 +21,25 @@
...
+ *
  * @ORM\Table(name="ohrm_performance_tracker_log")
  * @ORM\Entity
  */
 class PerformanceTrackerLog
 {
+    use DecoratorTrait;
+
+    public const POSITIVE_RATING = 1;
+    public const NEGATIVE_RATING = 2;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPerformancePlugin/entity/PerformanceTrackerLog.php:38]:
```suggestion
    public const POSITIVE_ACHIEVEMENT = 1;
    public const NEGATIVE_ACHIEVEMENT = 2;
```

File: src/plugins/orangehrmPerformancePlugin/entity/PerformanceTrackerLog.php [hunk 7]
```diff
@@ -42,78 +54,69 @@ class PerformanceTrackerLog
...
-     *   @ORM\JoinColumn(name="performance_track_id", referencedColumnName="id")
-     * })
+     * @ORM\ManyToOne (targetEntity="OrangeHRM\Entity\PerformanceTracker")
+     * @ORM\JoinColumn(name="performance_track_id", referencedColumnName="id")
+     *
      */
     private ?PerformanceTracker $performanceTracker;
 
     /**
      * @var Employee|null
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPerformancePlugin/entity/PerformanceTrackerLog.php:106]: are you sure these fields nullable

File: src/plugins/orangehrmPerformancePlugin/Api/Traits/PerformanceTrackerPermissionTrait.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    protected function checkTrackerLogEditable(PerformanceTrackerLog $performanceTrackerLog): bool
+    {
+        return $this->getTrackerLogPermission($performanceTrackerLog)->canUpdate();
+    }
+
+    /**
+     * @param PerformanceTrackerLog $performanceTrackerLog
+     * @return ResourcePermission
+     */
+    private function getTrackerLogPermission(PerformanceTrackerLog $performanceTrackerLog): ResourcePermission
```
[the remark this answers is not in the document]
Comment (SharukaNP) [reply src/plugins/orangehrmPerformancePlugin/Api/Traits/PerformanceTrackerPermissionTrait.php:45]: Can keep as protected right?

File: src/plugins/orangehrmPerformancePlugin/entity/PerformanceTrackerLog.php (hunk 7 above)
Comment (SharukaNP) [reply src/plugins/orangehrmPerformancePlugin/entity/PerformanceTrackerLog.php:106]: It is weird but except for id all fields are default NULL

**Review by devishke-orange — CHANGES_REQUESTED** [review]
few changes :+1: mostly formatting stuff

File: installer/Migration/V5_1_0/permission/data_group.yaml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+performance_tracker_log:
+  description: 'Performance - Tracker log'
+  allowed:
+    read: true
+    create: true
+    update: true
+    delete: true
+  permissions:
+    - { role: Admin, permission: { read: true, create: true, update: true, delete: true, self: false } }
+    - { role: Admin, permission: { read: true, create: true, update: true, delete: true, self: true } }
```
[the remark this answers is not in the document]
Comment (SharukaNP) [reply installer/Migration/V5_1_0/permission/data_group.yaml:10]: Need to allow admin to crud their own logs

File: src/plugins/orangehrmPerformancePlugin/entity/PerformanceTrackerLog.php (hunk 7 above)
Comment (RajithaKumara) [reply src/plugins/orangehrmPerformancePlugin/entity/PerformanceTrackerLog.php:106]: shall we make not nullable at entity level
```suggestion
     * @var Employee
```
and change the PHP type `?Employee` ==> `Employee`

same for `$user` field also

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**